### ISSUE TITLE
Feature request: opt-in auctionYield to yield the main thread between bid adapters in callBids

### ISSUE THREAD

**Issue by marco-prontera** [prebid/Prebid.js#15625]
I am working to reduce the main-thread cost of our third-party script, which runs Prebid.js on a number
of publisher sites. This proposal concerns a problem already raised here.

`adapterManager.callBids` calls every client bid adapter inside a single synchronous `forEach`. Each
iteration runs that adapter's `buildRequests` and dispatches its HTTP request, and nothing in between
returns control to the browser. On a site with many adapters that is one long task — I measured
**79–125 ms** on a live publisher page — and it is the largest uninterruptible block Prebid produces. 
An interaction landing in that window cannot be handled until it ends, which is what INP measures.

The problem has been raised twice and stopped for the same reason both times. Quoting the threads so
the history is in one place:

- **#5054 "Split long task"** (2020) — a publisher profiling Prebid found it shows up as a long task,
  "particularly with a lot of bidders". Reporting back on that investigation in #10062, @nanek was
  specific about where the time went: _"the majority of the time was in `callBids` for my site with 10+
  bidders and multiple ad units on the page. Breaking up `callBids` alone split up the long task into
  many smaller tasks."_
- **#10062** — the general "yield the main thread" request. @dgirardi: _"Prebid is actually
  intentionally hoarding the main thread […] I lost that argument at the time but I am still of the
  opinion that we should give up control."_ The counter-argument, from @snapwich: _"the whole concept
  of header bidding is to hold an auction in the browser before the page loads […] anytime we defer
  execution we work against that goal."_ @dgirardi later: _"periodically giving up control is a good
  idea (and, in fact, always better UX), but I'm in the minority."_
- In March 2024, @Deimos01 asked for exactly this: _"I would like to yield `callBids` in order to speed
  up INP metric."_
- **#12025** ("Core: start yielding control of the main thread", merged 2025-03-01) shipped the escape
  hatches — `GreedyPromise` became optional — plus `pbjs.yield` for the command queue and `renderAd`.

What follows is a proposal that I believe serves those publishers without changing anything for anyone who does not ask for it.

## Why the existing levers cannot do this

I tried both before proposing anything new. Neither works, and the reason is structural rather than a matter of tuning.

### `pbjs.yield`

Scoped to `_processQueue` (the `cmd`/`que` queue) and `renderAdDirect`. It never runs on the auction path.

### Opting out of `GreedyPromise`

This is the "quick test" suggested in #10062: make `PbPromise` a real `Promise`, so Prebid's existing
promises stop resolving synchronously and start breaking up tasks.

It does not split this loop, because **there is no promise on this path to make asynchronous**:

- `clientBidderRequests.forEach` in `src/adapterManager.ts` is plain synchronous iteration.
- `fun-hooks` dispatches `async` hooks through **callbacks, not promises** (there is no `Promise` in
  `fun-hooks/no-eval/index.js` at all), so `processBidderRequests` adds none.
- The promises in `src/ajax.ts` are on the **response** side — `fetchPm.then(...)`, after the request
  has already been sent.
- Of the places in core that do construct a `PbPromise` — `debugging`, `consentHandler`, `adRendering`,
  `auction.ts:623`, `prebid.ts:804`, `secureCreatives`, `utils.js:340`, `creativeRenderers`, `fpd/*` —
  none sits between `adapter.callBids` and `spec.buildRequests`.

I verified this on the page rather than assuming it. Same page, same bundle, same browser session,
4× CPU throttling, with `pbjs.Promise = window.Promise` seeded before Prebid loads (confirmed active at
runtime):

| `callBids` loop span                       | default (`GreedyPromise`) | `pbjs.Promise = window.Promise` |
| ------------------------------------------ | ------------------------- | ------------------------------- |
| 10 bidders                                 | 67.9 ms                   | 84.6 ms                         |
| 13 bidders                                 | 57.0 ms                   | 64.1 ms                         |
| 10 bidders                                 | 29.1 ms                   | 36.8 ms                         |
| longest task over the auction-start window | 94 · 77 · 93 ms           | 117 · 86 · 50 ms                |

No improvement — marginally worse, though at that scale it tracks whatever else the thread is doing —
and the long task over the auction start survives intact. Opting out of `GreedyPromise` breaks up
plenty of _other_ Prebid work; it does not touch this loop.

The same argument rules out a variant I considered: a custom `pbjs.Promise` whose `then` yields via
`scheduler.yield()`. It would still only intercept code that already builds a `PbPromise`, and this
path builds none.

### The workaround I tried: hooking `processBidderRequests`

Since neither lever works, I built a **workaround**. `callBids` cannot be modified from outside Prebid,
but there is one seam I can reach: the named `async` hook `processBidderRequests`, which Prebid invokes
once per bidder, immediately before that adapter's `buildRequests`. Yielding in a `before` hook there
lifts each `buildRequests` call — and the ajax dispatch after it — out of the loop's task and into a
task of its own.

That is the whole of it:

```js
getHook('processBidderRequests').before((next, ...args) => {
	const wrapCallback = args[4]; // config.callbackWithBidder(<bidderCode>)
	const callbacks = args[5]; // { onRequest, onBid, onCompletion, ... }

	// resume inside the bidder's config context; complete the bidder if it throws
	const resume = wrapCallback(() => {
		try {
			next(...args);
		} catch (e) {
			callbacks.onCompletion();
			throw e;
		}
	});

	typeof scheduler?.yield === 'function' ? scheduler.yield().then(resume, resume) : resume();
});
```

It splits the `buildRequests` calls, and section 3 measures how much that buys. But it reaches the
right work through the wrong door, and the door limits what it can do. Both halves are worth spelling
out, because together they are the argument for putting the yield in core.

**Why it works at all.** An `async` `fun-hooks` hook is not a promise — it is a **callback boundary
that can be left unresolved**. The chain walker hands each `before` hook a `next` continuation and does
not care when it is invoked:

```js
// fun-hooks/no-eval/index.js
function next(value) {
	if (localOrder[curr]) {
		var args = rest(arguments);
		args.unshift(next); // `next` handed to the before-hook as its 1st arg
		return localOrder[curr++].apply(thisArg, args);
	}
	...
}

localOrder[targetIndex] = function () {
	var result = target.apply(thisArg, args); // the real processBidderRequests body
	if (type === "sync") { next(result); }    // only 'sync' auto-advances
};

next.apply(null, args);
return result; // for 'async', `result` is never assigned — the trap returns undefined immediately
```

Two consequences, pulling in opposite directions.

1. **A `before` hook that does not call `next` synchronously suspends the hooked function.** The target
   body — `buildRequests` plus the ajax dispatch — stays captured in the closure until `next` runs, so a
   yield there moves that work into a later task. This is legal precisely because the hook is typed
   `async`; `src/hook.ts` documents the converse: _"you must not call `next` asynchronously from 'sync'
   hooks"_. It is what makes the workaround possible.
2. **But the caller does not wait.** For an `async` hook, `result` is only ever assigned in the
   `type === "sync"` branches, so the trap returns `undefined` as soon as the synchronous part of the
   chain unwinds. `adapter.callBids` returns, and `adapterManager`'s `forEach` — which ignores the
   return value — moves straight to the next bidder, in the same task.

**Point 2 is why this is a workaround and not a fix: the loop never pauses.** What actually happens:

```
task 1   the whole forEach, uninterrupted:
           bidder 1..N: guardTids, validate, ajax builder, BID_REQUESTED, hook -> queue a continuation
              ↓ N continuations queued
task 2     bidder 1: buildRequests + ajax dispatch
task 3     bidder 2: buildRequests + ajax dispatch
...
task N+1   bidder N: buildRequests + ajax dispatch
```

Every `buildRequests` ends up in its own task, which is the win. But the loop's per-iteration cost —
`guardTids`, bid validation, the ajax builder, emitting `BID_REQUESTED` — is still one uninterrupted
block, because the `forEach` completes before any continuation fires. Section 3 measures that residue at
**~24 ms for 17 bidders**.

A yield placed _inside_ the loop, where the `forEach` genuinely stops between bidders, has no such
split personality. It takes the per-iteration cost too, and needs no reliance on argument positions or
hook-dispatch internals to do it. That is Prebid's own code, and it is what this issue asks for.

## Evidence

All numbers come from one live publisher page: Prebid **11.26.0**, desktop Chrome, **4× CPU
throttling**, roughly 6 auctions per configuration. I throttle because on an unthrottled desktop most of
these blocks fall under the 50 ms long-task threshold and are invisible; 4× approximates the mid-tier
mobile hardware where INP problems occur.

**How I measured, so you can reproduce it without instrumenting a build.** Prebid already records what
is needed. `pbjs.getEvents()` returns every event since page load, each with an `elapsedTime` — a
`performance.now()` timestamp taken at emit. I correlate that log with a `PerformanceObserver` for
`longtask` and `long-animation-frame`. Three mappings make the log readable as a profile:

- `auctionInit` is emitted after `makeBidRequests` and immediately before `adapterManager.callBids`, so
  `auctionInit − beforeRequestBids` is the auction-start preamble.
- `bidRequested` is emitted once per bidder, immediately before that adapter's `callBids`. The gap
  between consecutive marks is that adapter's whole cost; the span from first to last is what I call the
  **`callBids` loop span** throughout this issue.
- `beforeBidderHttp` is emitted inside `processBidderRequests`, after `buildRequests` and before the
  ajax call.

Per-adapter internals come from Prebid's own metrics, attached to every bid: `bid.metrics.getMetrics()`
exposes `adapter.client.buildRequests` and `adapter.client.validate`.

"**Long task over the auction-start window**" means the longest `longtask` entry overlapping the
interval from `auctionInit` to the last `bidRequested`. `0` means no task of 50 ms or more overlapped
that window at all.

### 1. The loop is one long task, and `buildRequests` is what fills it

|                                                | value                                         |
| ---------------------------------------------- | --------------------------------------------- |
| single task containing the whole auction start | **79–125 ms** (5 auctions, 10–17 bidders)     |
| of which: `makeBidRequests` + RTD/floors hooks | 13–19 ms                                      |
| of which: the `callBids` loop                  | **57–78 ms (~70–75%)**                        |
| per adapter, `buildRequests`                   | median 1 ms (appnexus 4.5 ms, PBS s2s 8–9 ms) |
| per adapter, `isBidRequestValid`               | ~0 ms in every sample                         |

No single adapter is slow. The block is many small ones with no interruption between them — the shape a
yield fixes and adapter-level optimization does not.

### 2. Yielding with `scheduler.yield()` costs nothing; a timer costs a lot

This is the measurement I think should reopen the decision: a microbenchmark on the same page under the
same conditions, 17 iterations of 4 ms of synthetic work — the measured shape of the loop — with
different things between iterations, 3 runs each.

The figure is total wall-clock for the loop, which is **the delay added to the last bidder's HTTP
request**. That is the quantity the 2023 objection was about.

| between iterations                              | total loop                               |
| ----------------------------------------------- | ---------------------------------------- |
| nothing (today)                                 | 72, 80, 74 ms — one uninterruptible task |
| `scheduler.yield()`                             | **75, 72, 73 ms**                        |
| `scheduler.postTask({priority:'user-visible'})` | 85, 79, 75 ms                            |
| `setTimeout(0)`                                 | **115, 142, 150 ms**                     |

`scheduler.yield()` resumes at the front of the task queue, so the work finishes in the same wall-clock
time while becoming interruptible between steps. `setTimeout(0)` sends the continuation to the back of
the queue and costs +50–70 ms here — more on an ad-heavy page competing with third-party tasks of
200–500 ms.

In 2023 the only options were timers, so "deferring works against getting ads rendered with the page"
was sound. With `scheduler.yield()` the premise no longer holds.

### 3. A/B of the workaround on the real page

This measures the `processBidderRequests` workaround against the same build without it. Both bundles
were built from the same commit, differing only by that hook, and served to the same page in the same
browser session by overriding the script content — so network, layout, consent state and third parties
are identical.

|                                                     | baseline        | with yield         | Δ                                      |
| --------------------------------------------------- | --------------- | ------------------ | -------------------------------------- |
| `callBids` loop span (17 bidders, median)           | 65.0 ms         | **24.1 ms**        | **−63%**                               |
| loop span, all samples                              | 27–72 ms        | 8–26 ms            |                                        |
| long task containing the auction start (17 bidders) | 92 · 78 · 84 ms | 0 · 51 · 0 · 60 ms | no long task at all in 2 of 4 auctions |
| **total auction duration (median)**                 | **979 ms**      | **892 ms**         | no regression                          |
| bidders timed out                                   | 0               | 0                  | —                                      |

Total auction duration is measured from the first `bidRequested` to `auctionEnd`. It is
network-dominated and noisy, so the honest reading is "no measurable regression", not "an improvement" —
but it is the row that matters for the 2023 objection, and it does not move.

**The residual 24 ms is the loop itself**: deferring `next` does not pause the `forEach`, so only the
`buildRequests` calls move out. A yield inside `callBids` would take the rest too. **These numbers are
therefore a lower bound on what the in-core change achieves** — the workaround is the floor, not the
ceiling.

To be clear, I am not holding that hook up as a pattern to copy. It leans on argument positions and
hook-dispatch internals that carry no stability guarantee, which is exactly why I would rather the yield
lived in core.

### Limits of this evidence

I measured main-thread behavior, not revenue. I cannot claim a revenue-neutral result from this data —
only that auction duration and timeout counts did not move.

## Proposal

Add a global flag, read the same way `pbjs.yield` is read:

```js
pbjs.auctionYield = true; // default: false
```

and use Prebid's existing `yieldAll` from `src/utils/yield.ts` for the client-adapter loop in
`adapterManager.callBids`:

```js
// src/adapterManager.ts — today
clientBidderRequests.forEach((bidderRequest) => {
	/* guardTids, validate, ajax builder, BID_REQUESTED, adapter.callBids */
});

// proposed
yieldAll(
	() => getGlobal().auctionYield ?? false,
	clientBidderRequests.map((bidderRequest) => () => callClientAdapter(bidderRequest))
);
```

`doYield()` in `src/utils/yield.ts` already falls back to `PbPromise.resolve()` when `scheduler.yield`
is unavailable, which under the default `FEATURES.GREEDY` build resolves synchronously — no yield at
all, rather than a timer-based one. Given section 2, I think that conservative fallback is right here
and worth an explicit comment so it is not "fixed" later.

### Why a new flag rather than extending `pbjs.yield`

`pbjs.yield` defaults to `true` and is documented as covering command-queue processing. Extending its
scope to the auction would change auction behavior for every publisher on upgrade, unasked — precisely
the risk raised in #10062. A separate `auctionYield`, defaulting to `false`:

- is a **no-op for every existing publisher** until they opt in;
- keeps the two concerns independently switchable;
- lets publishers who want it A/B against their own revenue metrics before committing.

Shipping it as experimental also keeps the door open both ways: it can be dropped, or documented and
kept, based on the impact publishers report here.

### Why it matters

When `requestBids` runs, and how quickly those requests go out, matters: the sooner an auction starts,
the sooner the slots can be sold. But that has to coexist with interactivity. Publishers want a page
that stays responsive while the auction is in flight, because that is what keeps people on the site.

`scheduler.yield()` is what makes both possible. The loop becomes interruptible, and because the
continuation resumes at the front of the queue rather than the back, the requests still leave at
essentially the same time — section 2 measures that cost at zero. I would like to see this in core, with
publishers free to turn it on and report back what they measure.

## Happy to help

I can A/B the real in-core implementation on the same traffic and report loop span, long-task
distribution, auction duration, timeout rate and revenue. If it helps, I can open the PR.

## References

**Prebid**

- #10062 — Interest on Yielding on the Main Thread: https://github.com/prebid/Prebid.js/issues/10062
- #5054 "Split long task" (2020): https://github.com/prebid/Prebid.js/issues/5054
- #12025 — Core: start yielding control of the main thread: https://github.com/prebid/Prebid.js/pull/12025
- `pbjs.yield` docs (current scope): https://docs.prebid.org/dev-docs/publisher-api-reference/yield.html
- Source (11.26.0): `src/adapterManager.ts` (`callBids`), `src/utils/yield.ts` (`pbYield`, `yieldAll`,
  `doYield`), `src/adapters/bidderFactory.ts` (`processBidderRequests`), `src/hook.ts`

**Platform**

- `scheduler.yield()` shipped in Chrome 129, Sept 2024: https://chromestatus.com/feature/6266249336586240
- WICG explainer, yield and continuation: https://github.com/WICG/scheduling-apis/blob/main/explainers/yield-and-continuation.md
- Optimize long tasks: https://web.dev/optimize-long-tasks
- Long Animation Frames API (used for the attribution above): https://developer.chrome.com/docs/web-platform/long-animation-frames

**Comment by patmmccann** [comment]
this is very interesting. are you on the chrome team? I just got a similar ping from them today

>Opting out of GreedyPromise
>This is the "quick test" suggested in https://github.com/prebid/Prebid.js/issues/10062: make PbPromise a real Promise, so Prebid's existing
promises stop resolving synchronously and start breaking up tasks.

This test should already be how things work since https://github.com/prebid/Prebid.js/pull/12025 so I think I'm confused on what you tested there.

Adding the yield flag and https://github.com/prebid/Prebid.js/pull/15421 are the result of revenue concerns associated with yielding. Prebid 11 consistently gets about 1/2 point revenue loss in testers from 10 and https://github.com/prebid/Prebid.js/pull/15421 is the intended solve. 

The question on the yield stuff has always been where to put them; your suggestion is very helpful

Are you coming to prebid devcon next week? maybe we can talk in person

**Comment by patmmccann** [comment]
cc  @gilbertococchi who mentioned the successful yielding on raptive sites earlier today

**Comment by marco-prontera** [comment]
Hi @patmmccann! I wish, but no, I’m not on the Chrome team.

By way of introduction, I’m a software engineer at Talks Media. I initially developed our CMP, and after shipping its first release I began focusing more deeply on web performance and consent-related infrastructure. That work also led me to contribute improvements to the IAB Tech Lab’s iabtcf-es library, including this one: https://github.com/InteractiveAdvertisingBureau/iabtcf-es/pull/407

More recently, I’ve been working on our wrapper script, which uses Prebid.js. That brought me deeper into Prebid’s internals, and I’m currently analyzing its main-thread performance and its impact on UX and Core Web Vitals, particularly INP.

Regarding GreedyPromise, I tested Prebid 11.26.0 both with the default implementation and with an explicit opt-out by setting pbjs.Promise = window.Promise before Prebid loaded, then confirming the replacement at runtime. My point was not that opting out is unavailable, but that it does not break up this particular section: the client-adapter loop in `callBids` remains synchronous and does not create a `PbPromise` between bidders. The native Promise does split other parts of Prebid’s work, but it cannot introduce an asynchronous boundary along this specific path.

I understand the revenue concern, and I’m going to take a closer look at #15421 as well. My intention with the separate opt-in flag was precisely to make the behavior testable by individual publishers without changing the default auction path.

Unfortunately, I won’t be at Prebid DevCon next week, but I’d be very happy to continue the discussion here or arrange a call. I hope to be present in the future, for me this it's a awesome job.
